sycl/doc/extensions/experimental/sycl_ext_oneapi_kernel_properties.asciidoc

```diff
@@ -211,6 +235,18 @@ template <> struct is_property_key<device_has_key> : std::true_type {};
  of the work-group used to invoke the kernel. The order of the template
  arguments matches the constructor of the `range` class.
 
+|`max_work_group_size`
+|The `max_work_group_size` property provides a promise to the compiler
+that the kernel will never be launched with a larger work-group than the
+specified size. The number of template arguments in the `Dims` parameter pack
+must match the dimensionality of the work-group used to invoke the kernel. The
+order of the template arguments matches the constructor of the `range` class.
```

Comment:
Should we document which exception is thrown when runtime-provided work-group size is bigger than one specified by the property? The same applies to the other property being added.

Reply:
Yes, we probably should. I think we just want to copy the exception behaviour from `reqd_work_group_size`:

``` quote
If the kernel is submitted to a device that does support the work-group size,
but the application provides an nd_range that does not match the size from
the attribute, then the implementation must throw a synchronous exception with the
errc::nd_range error code.
```

Unfortunately it doesn't seem like we're actually doing this, at least when compiling SYCL for CUDA - even for the attribute.

Reply:
I've added wording to this effect. @gmlueck do you want to take another look?

Reply:
The specification looks good, but you say in the comment above that the implementation doesn't actually do this.  Can you add that support?

Reply:
Oh ... we document that these kernel properties work for all backends.  Do we have e-2-e tests that are enabled for all backends?  Those tests should also check for the exception (as we do in the test you reference above).

I see what you mean about the existing problem that CUDA and HIP don't generate the exception.  Do we have a tracker that captures this?  In general, we shouldn't disable tests unless we have a tracker that captures the work to fix the problem.

Can we:

* Write tests for these new properties, including tests that check the exception
* Disable the exception tests on CUDA / HIP with a comment that this isn't supported yet
* Make sure we have a tracker to implement the exception
* Make time to resolve this tracker

Reply:
Yeah, seems like this issue has flown under the radar for a while - we weren't aware of the problem. I've created an internal ticket to fix the current bug for HIP/CUDA, and that work will presumably also fix these new properties too.

Regarding end-to-end testing for exceptions, I fear we might have backed ourselves into a corner:

* HIP/CUDA targets are broken for the aforementioned reasons
* The OpenCL CPU target doesn't support the `KernelAttributesINTEL` SPIR-V execution mode, so doesn't even support the `[[intel::max_work_group_size]]` _attribute_, let alone the kernel property.
* The OpenCL FPGA target does support that execution mode, but doesn't throw an exception, even for the attribute. Note how there are no end-to-end tests for the attribute either! This is arguably a bug with that target, but given that there's no specification for the attribute, it's not a firm footing. I think UR would need to receive this information from the Level Zero driver, and so we'd need to update the Level Zero specification to be able to provide those kernel properties.
* I haven't been able to test any other targets.

We would probably also need to extend `SPV_INTEL_kernel_attributes` to support the `max_linear_work_group_size` execution mode. Otherwise the information is lost. But then again the OpenCL CPU target doesn't support that extension.

If we do want to have these attributes in a specification, we definitely want to specify that they can throw an exception - otherwise they're useless. I don't think "can" is good enough either: we want guarantees. It just so happens that we're unable to meaningfully test this behaviour on any target right now.

Ironically it seems like HIP/CUDA might be the easiest to get working as they don't use SPIR-V.

Do you have any suggestions on how to proceed? It's an experimental extension but I don't know if that means we can half implement it. Can we selectively implement it for different targets? If the OpenCL CPU driver is never going to support the required SPIR-V mode, there's not much we can do.

Reply:
Note that #14634 re-enables those tests for HIP/CUDA so the fundamental issue should be fixed.

Reply:
@gmlueck how should we proceed with this issue? It's my opinion that this extension can't be supported by all targets (Intel targets), at least not in this PR. There are far deeper issues going on with both of those targets, least of all no support in SPIR-V for one of the properties.

We can make NVPTX/AMD support the property, but it'd involve a separate change to UR and for us to bump UR. Should that be part of this PR, or done as a follow-up?

Reply:
Sorry, can you summarize the current state on the various backends?  Are these two new properties supported on *any* backend currently?

Also, two questions regarding support on Intel devices via SPV:

* You mention above that the OpenCL CPU backend doesn't support the `KernelAttributesINTEL` execution mode.  What happens if you submit a module to that backend with this attribute?  Is it ignored, or does it raise an error?

* What is the support on the Level Zero and OpenCL GPU backends?

Reply:
> Sorry, can you summarize the current state on the various backends? Are these two new properties supported on _any_ backend currently?
> 
> Also, two questions regarding support on Intel devices via SPV:
> 
>     * You mention above that the OpenCL CPU backend doesn't support the `KernelAttributesINTEL` execution mode.  What happens if you submit a module to that backend with this attribute?  Is it ignored, or does it raise an error?
> 
>     * What is the support on the Level Zero and OpenCL GPU backends?

These properties are not fully supported on any backend, currently. The CUDA backend uses it for what it needs it for - to emit NVVM annotations - but no backend supports the exceptions:

CUDA, HIP, and Native CPU would need the SYCL runtime to query the UR adapters to receive the max work-group size information, and throw the right exception. This would require an update to the UR spec.

OpenCL CPU reports:
```
# | Build program log for '11th Gen Intel(R) Core(TM) i7-1165G7 @ 2.80GHz':
# | Compilation started
# | Unsupported SPIR-V module
# | KernelAttributes is only supported on FPGA emulator
# | Compilation failed
```
and crashes.

I don't really know how OpenCL GPU works. Presumably it uses SPIR-V, but would also need the change to UR to query for the program information. Note also that SPIR-V can only encode `MaxWorkGroupSize` but not "MaxLinearWorkGroupSize" so we'd need a new SPIR-V extension, or to update the `KernelAttributesINTEL` extension, before we can report an exception for SPIR-V programs. Either way, all of our SPIR-V consumers would need updating.

I also don't fully understand the Level Zero backend either. If it uses SPIR-V it will need the same spec changes as for OpenCL GPU. Regardless it would need the SYCL runtime to query UR (as before) which would then need to query Level Zero for the program information. This would require a change to the L0 specification.

I think that summarizes the current situation. Adhering to the exception behaviour would require at least one specification change for every backend (UR being essential for all of them), and I don't see an easy way for the OpenCL CPU given its behaviour with the required SPIR-V extension, and the lack of SPIR-V/LevelZero support for the "linear" property.

Maybe it's a chicken and egg situation, because to add these properties to UR we'd want to be reasonably assured that all backends _can_ support them, at least in principle.

Reply:
@gmlueck I've preemptively added these two properties to UR in https://github.com/oneapi-src/unified-runtime/pull/1996, and supported them for CUDA and Native CPU. The other adapters are marked as XFAIL, pending changes to SPIR-V/OpenCL/Level Zero specs. That might be enough to get this PR through the door?

Reply:
Let's do this.  Add a new section to the specification titled "Kernel Properties for the CUDA backend", or you can title it "Kernel Properties for the CUDA and native CPU backends" if you think the native cpu support is interesting.  This section should immediately follow "Kernel properties", and it should have an introduction that says something like:

> The kernel properties specified in this section may only be used to decorate kernels that are submitted to the CUDA backend.  Attempting to submit a kernel with these properties to another backend has undefined behavior.

Then move the synopsis and description of these two properties to that new section.

Do not include the paragraph saying that an exception is raised since that is not the case.  However, do add a note like this:

> [_Note:_ In the future, implementations may throw a synchronous exception with the `errc::nd_range` error code if the kernel is submitted with an `nd_range` that exceeds the size specified by the property.  However, that is not currently the case.  Instead, the behavior is undefined if the size of the `nd_range` exceeds the size of the property.
_{endnote}_]

(See [here](https://github.com/intel/llvm/blob/sycl/sycl/doc/extensions/proposed/sycl_ext_oneapi_work_group_memory.asciidoc) for an example of this "note" syntax.)

Reply:
Okay, that sounds good. But I should say (I maybe didn't make this clear) that the exception *is* currently raised for CUDA and Native CPU. It could be enabled for HIP with minimal effort, but we're currently not prioritising that adapter. If that would help come up with a better rationale than "Kernel Properties for the CUDA and Native CPU and HIP backends" then it might be worthwhile doing the work for HIP too.

I don't know if that changes your suggestion about moving these properties to a new section or not, but I would imagine we do want to keep the wording on exception behaviour.

Reply:
OK, I didn't realize that CUDA and native CPU implemented the exception.  In that case, we should still create the separate section, and also keep the wording about the exception.  The "note" is not needed.

Reply:
Great, thank you. That's done now.